tinygrad/tensor.py

```diff
@@ -380,28 +380,42 @@ def __getitem__(self, indices: Union[int, slice, Tensor, None, List, Tuple]) ->
     if type_dim[Tensor]:
 
       # extract tensors and tensor dimensions
-      idx, tdim = [], []
+      idx, tdim, empty_idx = [], [], False
       for tensor_dim in type_dim[Tensor]:
         dims_collapsed_, dims_injected = sum(1 for d in dims_collapsed if tensor_dim >= d), sum(1 for d in type_dim[None] if tensor_dim >= d)
         tdim.append(td := tensor_dim - dims_collapsed_ + dims_injected)
-        # normalize the negative tensor indices
         idx.append(((t := indices[tensor_dim + dims_injected]) < 0).where(ret.shape[td], 0) + t)
+        if 0 in t.shape: empty_idx = True
         # TODO uint8 and bool tensor indexing
         if not (dtypes.is_int(t.dtype) or t.dtype == dtypes.bool): raise IndexError("tensors used as indices must be int or bool tensors")
 
+      if empty_idx and 0 in ret.shape: return Tensor.empty(0, dtype=ret.dtype)
+
+      # manual broadcasting
+      final_shape = (shapes := tuple(list(i.shape) for i in idx))[0]
+      for interim_shape in shapes[1:]:
+        if len(final_shape) < len(interim_shape): final_shape = [1]*(len(interim_shape) - len(final_shape)) + final_shape
```

Comment:
can you refactor this with the logic in `_broadcasted`? the left pad with 1 and max from zip are the same